pkg/api/helpers.go

```diff
@@ -496,37 +496,81 @@ func GetTaintsFromNodeAnnotations(annotations map[string]string) ([]Taint, error
 }
 
 // TolerationToleratesTaint checks if the toleration tolerates the taint.
-func TolerationToleratesTaint(toleration *Toleration, taint *Taint) bool {
-	if len(toleration.Effect) != 0 && toleration.Effect != taint.Effect {
+func (t *Toleration) ToleratesTaint(taint *Taint) bool {
+	if len(t.Effect) != 0 && t.Effect != taint.Effect {
 		return false
 	}
 
-	if toleration.Key != taint.Key {
+	if t.Key != taint.Key {
 		return false
 	}
-	// TODO: Use proper defaulting when Toleration becomes a field of PodSpec
-	if (len(toleration.Operator) == 0 || toleration.Operator == TolerationOpEqual) && toleration.Value == taint.Value {
-		return true
+
+	// check forgivenessSeconds time out
+	if t.ForgivenessSeconds != nil {
+		// taint with no added time indicated can only be tolerated
+		// by toleration with no forgivenessSeconds.
+		if taint.AddedTime.IsZero() {
+			return false
+		}
+		if unversioned.Now().After(taint.AddedTime.Add(time.Second * time.Duration(*t.ForgivenessSeconds))) {
```

Comment:
I'm not sure what you mean - how NC currently works in HA depends strongly on how it is implemented. NC uses internal cache that uses internal clock with the evictions, so even if the "mastership" will be transferred, the behavior will be consistent with one of NC clocks. But that's the issue here - IIUC this is the skew that will happen all of the time, between master machine and kubelets. I'm not sure we care, but it's worth noting that it may cause problems, especially in not-sychcronized clusters.